### PR TITLE
Redact secrets when logging config and warn user when changing categories

### LINKED ISSUE #19465: Add user warning that logging can store PII on the user's machine
### Is your feature request related to a problem? Please describe.

NVDA is designed to operate locally without collecting user data. However, NVDA does provide logging capabilities and due to its privileged access to the user’s machine 'Debug' or 'Input/Output' logging levels can capture potentially sensitive information (eg passwords, PII) on the user's machine. 

### Describe the solution you'd like

Selecting debugging levels above 'Info' should display a warning to the user regarding the storage of sensitive data on their machine. This warning should be consistent with what is written in the user guide. The user guide should cover the case of a user selecting elevated debugging levels through the CLI.

### Describe alternatives you've considered

None.

### Additional context

n/a

### PR DESCRIPTION
**Pull request by seanbudd**
### Link to issue number:
Closes #19465

### Summary of the issue:
Secrets stored in NVDA config are often unintentionally logged in debug mode by NVDA

### Description of user facing changes:
logging will attempt to redact secrets when the developer decides to sanitise risky log messages.
Added a new log level: secrets, to disable redactions for required debug logging.
Added a warning whenever selecting a log level below info.

### Description of developer facing changes:
A new `redactSecrets` parameter for logging, which searches for and replaces secrets in the log message.

### Description of development approach:
Use https://github.com/Yelp/detect-secrets

This pull request introduces secret redaction support in logging, ensuring that sensitive information is masked in log outputs when requested.
A new log level is added so you can view unredacted logs if needed.

**Secret Redaction in Logging**

* Added a `redactSecrets` parameter to the `Logger._log` method in `logHandler.py` that, when enabled, uses the `detect-secrets` library to scan and mask detected secrets in log messages.
* Updated logging calls in `source/config/__init__.py` to use `redactSecrets=True` when logging potentially sensitive configuration data.
* Updated the developer documentation to describe the new `redactSecrets` parameter and recommend its use for sensitive data.

**Dependency and Packaging Support**

* Added `detect-secrets` as a dependency in `pyproject.toml` and ensured all relevant submodules are included in frozen builds for dynamic plugin loading
* Include `multiprocessing` in bundle - needed for import, seems to functionally work?

### Testing strategy:
 - [x] Manual testing by putting a secret in my NVDA config. 
    - [x] installed copy from build
    - [x] source copy
- [x] Added unit tests for the new secret redaction logic in `tests/unit/test_logHandler.py`, covering both normal and edge cases.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

### CODE REVIEW

File: source/logHandler.py [hunk 1]
```diff
@@ -273,6 +288,12 @@ def _log(
 				"".join(traceback.format_list(stack_info)).rstrip(),
 			)
 
+		if redactSecrets:
+			from detect_secrets.core.scan import scan_line
+
+			for secret in scan_line(msg: str):
+				msg.replace(secret.secret_value, "****")
```
Comment (seanbudd) [inline source/logHandler.py]:
```suggestion
				msg = msg.replace(secret.secret_value, "****")
```

**Comment by seanbudd** [comment]
huh - seems like the module still isn't being included correctly

File: source/gui/settingsDialogs.py [hunk 2]
```diff
@@ -6276,6 +6319,10 @@ def makeSettings(self, sizer: wx.BoxSizer):
 			self._allowUsageStatsCheckBox.Value = False
 			self._allowUsageStatsCheckBox.Disable()
 
+	def onApply(self, evt):
+
+		super().onApply(evt)
+
```
Comment (seanbudd) [inline source/gui/settingsDialogs.py]:
```suggestion
```

File: source/gui/settingsDialogs.py [hunk 3]
```diff
@@ -6301,10 +6344,16 @@ def onSave(self):
...
+			updateLogLevel = selectedLogLevel != self._savedLogLevel and (
+				selectedLogLevel >= LoggingLevel.INFO or self._confirmLogLevelChange(selectedLogLevel)
 			)
-			logHandler.setLogLevelFromConfig()
+			if updateLogLevel:
+				config.conf["general"]["loggingLevel"] = logging.getLevelName(selectedLogLevel)
+				logHandler.setLogLevelFromConfig()
+				self._savedLogLevel = selectedLogLevel
+			else:
+				log.debug("User cancelled log level change, keeping original control value.")
```
Comment (seanbudd) [inline source/gui/settingsDialogs.py]:
```suggestion
```

File: user_docs/en/changes.md [hunk 4]
```diff
@@ -71,6 +71,11 @@ Please refer to [the developer guide](https://download.nvaccess.org/documentatio
   * uv to 0.11.7. (#19548, #19908, #19968)
   * Requests to 2.33.0. (#19877)
   * cryptography to 46.0.7. (#19877, #19968)
+* A new parameter `redactSecrets` has been added to logging functions e.g. `log.debug`. (#19966)
+  * When set to `True`, logging output will be sanitized to replace detected secrets with asterisks.
+  * This is set to `False` by default for performance purposes.
+  * It is encouraged to enable this when logging anything particularly sensitive e.g. clipboard content.
+  * Added a `SECRETS` logging level for cases where developers explicitly need debug logging without `redactSecrets` masking. (#19966)
```
Comment (seanbudd) [inline user_docs/en/changes.md:82]: should this go under general changes? I'm not sure if we want to advertise this to users or not

File: source/logHandler.py [hunk 5]
```diff
@@ -233,15 +236,30 @@ class Logger(logging.Logger):
...
+
+		:param level: The severity level of the log message.
+		:param msg: The log message, which may contain format specifiers that will be replaced by the values in `args`.
+		:param args: The arguments to be merged into `msg` using the `%` operator for string formatting.
+		:param exc_info: Exception information to be logged, defaults to None
+		:param extra: Additional information to be logged, defaults to None
+		:param codepath: The code path where the log was generated, defaults to None
+		:param activateLogViewer: Whether to activate the log viewer, defaults to False
+		:param stack_info: Stack information to be logged, defaults to None
+		:param redactSecrets: Whether to check for and redact secrets in the log message, defaults to False
```
Comment (CyrilleB79) [inline source/logHandler.py]: No need to re-specify defaults: there are already type hints.

```suggestion
		:param exc_info: Exception information to be logged
		:param extra: Additional information to be logged
		:param codepath: The code path where the log was generated
		:param activateLogViewer: Whether to activate the log viewer
		:param stack_info: Stack information to be logged
		:param redactSecrets: Whether to check for and redact secrets in the log message
```

File: source/logHandler.py [hunk 6]
```diff
@@ -273,7 +291,26 @@ def _log(
 				"".join(traceback.format_list(stack_info)).rstrip(),
 			)
 
-		res = super()._log(level, msg, args, exc_info, extra)
+		if redactSecrets and self.getEffectiveLevel() != self.SECRETS:
```
Comment (CyrilleB79) [inline source/logHandler.py]: Just in case there is a lower log level in the future, shouldn't we use `>` instead?

Please confirm that there is no special case with log level = 0 or -1 though before accepting this suggestion.
```suggestion
		if redactSecrets and self.getEffectiveLevel() > self.SECRETS:
```

Comment (seanbudd) [reply source/logHandler.py]: I think that makes sense

**Comment by seanbudd** [comment]
unit tests are failing

**Review by Qchristensen — APPROVED** [review]
Changes read well

**Review by SaschaCowley — APPROVED** [review]
